lib/spack/spack/bootstrap.py

```diff
@@ -837,3 +838,156 @@ def ensure_flake8_in_path_or_raise():
+        'patch': _missing_executable('patch', 'patch source code before building'),
+        'bash': _missing_executable('bash',  'use Spack compiler wrapper'),
+        'tar': _missing_executable('tar', 'manage code archives'),
+        'gzip': _missing_executable('gzip', 'compress/decompress code archives'),
+        'unzip': _missing_executable('unzip', 'compress/decompress code archives'),
+        'bzip2': _missing_executable('bzip2', 'compress/decompress code archives'),
+        'git': _missing_executable('git', 'fetch/manage git repositories')
+    }
+    if platform.system().lower() == 'linux':
+        _core_system_exes['xz'] = _missing_executable(
+            'xz', 'compress/decompress code archives'
+        )
+
+    # Executables that are not bootstrapped yet
+    result = [functools.partial(_system_executable, exe, msg)
+              for exe, msg in _core_system_exes.items()]
+    # Python modules
+    result += [
+        functools.partial(
+            _required_python_module, 'clingo', clingo_root_spec(),
+            'MISSING: "clingo" python module (required to concretize specs)'
+        )
+    ]
+    return result
+
+
+def _buildcache_requirements():
+    _buildcache_exes = {
+        'file': _missing_executable('file', 'analyze files for buildcaches'),
+        ('gpg2', 'gpg'): _missing_executable('gpg2', 'sign/verify buildcaches')
+    }
+    if platform.system().lower() == 'darwin':
+        _buildcache_exes['otool'] = _missing_executable('otool', 'relocate binaries')
+
+    # Executables that are not bootstrapped yet
+    result = [functools.partial(_system_executable, exe, msg)
+              for exe, msg in _buildcache_exes.items()]
+
+    if platform.system().lower() == 'linux':
+        result.append(functools.partial(
+            _required_executable, 'patchelf', patchelf_root_spec(),
+            _missing_executable('patchelf', 'relocate binaries')
+        ))
+
+    return result
+
+
+def _optional_requirements():
+    _optional_exes = {
+        'zstd': _missing_executable('zstd', 'compress/decompress code archives'),
+        'svn': _missing_executable('svn', 'manage subversion repositories'),
+        'hg': _missing_executable('hg', 'manage mercurial repositories')
+    }
+    # Executables that are not bootstrapped yet
+    result = [functools.partial(_system_executable, exe, msg)
+              for exe, msg in _optional_exes.items()]
+    return result
+
+
+def _development_requirements():
+    return [
+        functools.partial(
+            _required_executable, 'isort', isort_root_spec(),
+            'MISSING: "isort" executable (required for style checks)'
+        ),
+        functools.partial(
+            _required_executable, 'mypy', mypy_root_spec(),
+            'MISSING: "mypy" executable (required for style checks)'
+        ),
+        functools.partial(
+            _required_executable, 'flake8', flake8_root_spec(),
+            'MISSING: "flake8" executable (required for style checks)'
```

Comment:
Would it make sense to tell users how to install missing packages with Spack? e.g. in this case `spack install py-flake8` (in an environment).

Reply:
It will bootstrap itself automatically if you need it IIRC

Reply:
A few specs will bootstrap automatically, like this one as @becker33  said, others are system requirement. I think we might want an option or a subcommand that bootstraps whatever is missing in one shot, but I'd prefer to leave that as a future PR.

Right now it's not clear to me what would be the best interface. Perhaps:
```console
$ spack bootstrap status --fix
```
?

Reply:
In that case I'm worried that the user would think they needed to do something when in fact Spack would bootstrap the executable for them (perhaps the message should just say "spack will take care of this when you run any command that needs it", but in that case, I'm not sure if it makes sense to report on its status).

Reply:
@scheibelp Right now I mark with a `[B]` missing software that can be bootstrapped and with `[-]` missing software that must be on the system (see the updated image in the description). I can also add a closing sentence like:
```
Spack will take care of bootstrapping any missing dependency marked as [B]. Dependencies marked as [-] are instead  required to be found on the system.
```
if the command detects anything missing. Would that be fine?

Reply:
Yes that sounds great. In https://github.com/spack/spack/pull/28004#discussion_r773430218, I was questioning whether we would want to mention software that we are going to bootstrap anyway (if we are making it seamless, I wasn't sure  why we'd mention it here) but on second thought I think this is just to keep track of all the software we need outside of Spack.

Reply:
Done in d0b9dd7